transport-native-kqueue/src/main/java/io/netty/channel/kqueue/AbstractKQueueChannel.java

```diff
@@ -405,15 +405,8 @@ final void readReady(long numberBytesPending) {
 
         final void readReadyFinally(ChannelConfig config) {
             maybeMoreDataToRead = allocHandle.maybeMoreDataToRead();
-            // Check if there is a readPending which was not processed yet.
-            // This could be for two reasons:
-            // * The user called Channel.read() or ChannelHandlerContext.read() in channelRead(...) method
-            // * The user called Channel.read() or ChannelHandlerContext.read() in channelReadComplete(...) method
-            //
-            // See https://github.com/netty/netty/issues/2254
-            if (!readPending && !config.isAutoRead()) {
-                clearReadFilter0();
-            } else if (readPending && maybeMoreDataToRead) {
+
+            if ((readPending || allocHandle.isReadEOF()) && maybeMoreDataToRead) {
```

Comment:
Should this conditional instead be the following:

`(allocHandle.isReadEOF() || (readPending && maybeMoreDataToRead))`

This says: "if EOF we always drain all data, otherwise we only read if there maybe data pending and a read is requested"

This would allow us to remove the EOF check from the `maybeMoreDataToRead()` method ... which is now duplicated inside/outside the method. We can update the comments in the `maybeMoreDataToRead()` method to say "it is assumed EOF is handled externally" (in addition to other comments). Since we have to expose `isReadEOF()` now (and we already do this for EPOLL) it seems to make sense.